zdx/trailer.go

```diff
@@ -0,0 +1,178 @@
+package zdx
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/brimsec/zq/zcode"
+	"github.com/brimsec/zq/zio/zngio"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+)
+
+const (
+	MagicField       = "magic"
+	VersionField     = "version"
+	ChildField       = "child_field"
+	FrameThreshField = "frame_thresh"
+	SectionsField    = "sections"
+	KeysField        = "keys"
+
+	MagicVal      = "microindex"
+	VersionVal    = "0.3"
+	ChildFieldVal = "_child"
+
+	TrailerMaxSize = 4096
+)
+
+type Trailer struct {
+	Magic            string
+	Version          string
+	ChildOffsetField string
+	FrameThresh      int
+	KeyType          *zng.TypeRecord
+	Sections         []int64
+}
+
+var ErrNotIndex = errors.New("not a microindex")
+
+func newTrailerRecord(zctx *resolver.Context, childField string, frameThresh int, sections []int64, keyType *zng.TypeRecord) (*zng.Record, error) {
+	sectionsType := zctx.LookupTypeArray(zng.TypeInt64)
+	cols := []zng.Column{
+		{MagicField, zng.TypeString},
+		{VersionField, zng.TypeString},
+		{ChildField, zng.TypeString},
+		{FrameThreshField, zng.TypeInt32},
+		{SectionsField, sectionsType},
+		{KeysField, keyType},
+	}
+	typ, err := zctx.LookupTypeRecord(cols)
+	if err != nil {
+		return nil, err
+	}
+	builder := zng.NewBuilder(typ)
+	return builder.Build(
+		zng.EncodeString(MagicVal),
+		zng.EncodeString(VersionVal),
+		zng.EncodeString(childField),
+		zng.EncodeInt(int64(frameThresh)),
+		encodeSections(sections),
+		nil), nil
+}
+
+func encodeSections(sections []int64) zcode.Bytes {
+	var b zcode.Builder
+	for _, s := range sections {
+		b.AppendPrimitive(zng.EncodeInt(s))
+	}
+	return b.Bytes()
+}
+
+func readTrailer(r io.ReadSeeker, n int64) (*Trailer, int, error) {
+	if n > TrailerMaxSize {
+		n = TrailerMaxSize
+	}
+	if _, err := r.Seek(-n, io.SeekEnd); err != nil {
+		return nil, 0, err
+	}
+	buf := make([]byte, n)
+	cc, err := r.Read(buf)
+	if err != nil {
+		return nil, 0, err
+	}
+	if int64(cc) != n {
+		// this shouldn't happen but maybe could occur under a corner case
+		// or I/O problems XXX
+		return nil, 0, fmt.Errorf("couldn't read trailer: expected %d bytes but read %d", n, cc)
+	}
+	for off := int64(n) - 4; off >= 0; off-- {
+		// look for end of stream followed by an array[int64] typedef then
+		// a record typedef indicating the possible presence of the trailer,
+		// which we then try to decode.
+		if buf[off] == 0x85 && buf[off+1] == 0x81 && buf[off+2] == 0x06 && buf[off+3] == 0x80 {
+			attempt := buf[off+1 : n]
+			rec := readOne(bytes.NewReader(attempt))
+			if rec == nil {
+				continue
+			}
+			trailer, _ := recordToTrailer(rec)
+			if trailer != nil {
+				return trailer, len(attempt), nil
+			}
+		}
+	}
+	return nil, 0, errors.New("trailer not found")
+}
+
+func readOne(reader *bytes.Reader) *zng.Record {
+	r := zngio.NewReader(reader, resolver.NewContext())
+	rec, err := r.Read()
+	if err != nil && err != io.EOF {
+		rec = nil
+	}
```

Comment:
A zbuf.Reader.Read implementation should never io.EOF and should never return both a non-nil record and a non-nil error.
```suggestion
	rec, _ := r.Read()
```

Reply:
(If you've observed either behavior from zngio.Reader.Read, you've found a bug that I want to fix.)

Reply:
Cool, thanks.  As documented in the interface declaration :-)